feat: Scrub filenames in the minidump modules list

flub opened the pull request:
This allows selecting minidump module code and debug filepaths to be
scrubbed.  The basename of these filepaths are preserved in order to
ensure we can still process the minidumps.

Commit:
feat: Scrub filenames in the minidump modules list

The ValueType::Filepath is added and replaces NativeDebugImage, it
identifies a value which is a string containing a pathname.  This is
then used when scrubbing pathnames in the minidump module list.

flub commented on relay-general/src/pii/attachments.rs:
```diff
@@ -365,6 +365,13 @@ pub struct PiiAttachmentsProcessor<'a> {
 }
 
+/// Which encodings to scrub for `scrub_bytes`.
+pub enum ScrubEncodings {
```
I find this and especially it's use horrible :(

untitaker replied:
make it a struct of bools/bitset?

flub replied:
Not sure that's worth it for now.  What I really struggled with was to make `scrub_bytes()` nice to implement for all these combinations.  I just did it the dumb, duplicating way for now but it certainly doesn't make me happy.

untitaker replied:
I think it's worth it... I don't see much work in deduplicating it?

flub commented on relay-general/src/pii/attachments.rs:
```diff
@@ -455,7 +487,49 @@ impl<'a> PiiAttachmentsProcessor<'a> {
+
+    /// Scrub a filepath, preserving the basename.
+    pub fn scrub_utf8_filepath(&self, path: &mut str, state: &ProcessingState<'_>) -> bool {
+        eprintln!("utf8 path: {}", path);
```
Oops, still have to remove all these :)

flub commented on relay-general/src/protocol/debugmeta.rs:
```diff
@@ -23,21 +23,22 @@ use crate::types::{
 #[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
-pub struct NativeImagePath(pub String);
+pub struct Filepath(pub String);
+// pub struct Filepath(#[metastructure(value_type = "Filepath")] pub String);
```
@untitaker I'm not sure what the consequences are here, but this broke.

untitaker replied:
wups, fixed

untitaker commented on relay-general/src/pii/attachments.rs:
```diff
@@ -455,7 +487,49 @@ impl<'a> PiiAttachmentsProcessor<'a> {
+    pub fn scrub_utf8_filepath(&self, path: &mut str, state: &ProcessingState<'_>) -> bool {
+        eprintln!("utf8 path: {}", path);
+        if let Some(index) = path.rfind(|c| c == '/' || c == '\\') {
+            let data = unsafe { &mut path.as_bytes_mut()[..index] };
```
Why does this need to be unsafe?

flub replied:
`str::as_bytes_mut` is unsafe since it allows you to write invalid UTF-8 into a `str`

untitaker committed:
fix annotation

untitaker commented on relay-general/src/pii/minidumps.rs:
```diff
@@ -130,21 +240,64 @@ impl PiiAttachmentsProcessor<'_> {
-            changed |= self.scrub_bytes(slice, &state);
+            match item {
+                MinidumpItem::StackMemory(range) => {
+                    let slice = data
```
This stmt appears to be the same between all match arms

flub replied:
It's a consequence of me putting the range into the enum variants instead of carrying it outside of it.

untitaker commented on relay-general/src/pii/processor.rs:
```diff
@@ -100,13 +100,13 @@ impl<'a> Processor for PiiProcessor<'a> {
         self.apply_all_rules(meta, state, Some(value))
     }
 
-    fn process_native_image_path(
```
I think I regret this rename actually, can we go with nativeimagepath or imagepath all the way? The reason being that there's special behavior attached to this valuetype that makes no sense in other situations outside of filepaths that are relevant for processing.

@jan-auer wdyt? Intent is to let people type `$native_image_path` or something like that. Right now they write `debug_meta.images.*.code_file` which I think we should not emulate for minidump scrubbing

jan-auer replied:
I would definitely not expose this value type to data scrubbers, this should be fully embedded in processing. Otherwise, it breaks left and right if we have to evolve this later. 

What we **could** do is introduce a `$path` type, but then everything that could contain a path would probably have to be tagged by that. And I think we have attributes like `$frame.package` that can be either or, where it not necessarily makes sense to apply this logic. So I'd stay on the safer side and not do it. 

Particularly, the `$frame.package` attribute can contain a full absolute native image path. We cannot classify it as such, though, since it can also contain arbitrary other information in other platforms. 

Really, this is a matcher. Not a selector, and not a value type. We should have a rule where you can say "In [selector] if this matches a file location, then on the path apply [redaction]".

untitaker replied:
Ok you bring up some good points. I think I know what I want. I will push a solution and summarize later.

untitaker replied:
So to summarize:

* nativeimagepath stays as-is
* no valuetype, that means people will need to explicitly add `$attachments.*.code_file` (or just `code_file`)

flub commented on relay-general/src/pii/minidumps.rs:
```diff
@@ -130,21 +240,64 @@ impl PiiAttachmentsProcessor<'_> {
+                        Some(ValueType::Filepath),
+                    );
+                    let wstr = WStr::from_utf16le_mut(slice)?; // TODO: Consider making this lossy?
+                    changed |= self.scrub_utf16_filepath(wstr, &state);
```
```
"applications": {
    "code_file": ["anything:remove"]
}
``` 
currently still only gives me things like:
```
(code_file)                     = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx\crash.exe"
```
because `scrub_utf16_filepath` never tries to scrub the basename. Is that what we want?

flub replied:
i was expecting that for `"$filepath": ["anything:remove"]` but I'm probably misunderstanding selectors and ValueTypes

untitaker replied:
This lgtm. `$filepath` should behave the same (it's just an alias)

flub replied:
as i understand it it makes it impossible to completely remove these names as it will always be identified as this valuetype and only do partial scrubing.  as long as we want that, that's fine by me.

untitaker replied:
Yes we want that specifically because we don't want people to be able to break stacktrace processing with data scrubbing. Stacktrace processing needs the basename.

Commit:
Correctly replace debug names

Previously they were not being scrubbed.

Commit:
Support big- and little-endian minidumps

This uses the endian indicator in the minidump to achieve this.  It
does introduce a dependency on scroll directly as the minidump crate
does not expose the type otherwise.  We use the same version as the
minidump crate does.

Commit:
Add a test

This only tests a windows minidump so far.

untitaker committed:
revert changes made to nativeimagepath

jan-auer commented on relay-general/src/pii/attachments.rs:
```diff
@@ -455,7 +484,38 @@ impl<'a> PiiAttachmentsProcessor<'a> {
+    pub fn scrub_utf16_filepath(&self, path: &mut WStr, state: &ProcessingState<'_>) -> bool {
+        let mut found = false;
+        let mut index = 0;
+        for (i, c) in path.char_indices() {
```
It would be really nice to make this `DoubleEndedIterator` and then just `.reverse().find()`.

jan-auer commented on relay-general/src/pii/minidumps.rs:
```diff
@@ -130,23 +253,157 @@ impl PiiAttachmentsProcessor<'_> {
+                    let state =
+                        file_state.enter_static("code_file", Some(attrs), Some(ValueType::String));
+                    let wstr = WStr::from_utf16le_mut(slice)?; // TODO: Consider making this lossy?
+                    changed |= self.scrub_utf16_filepath(wstr, &state);
```
I think for these you should push a path onto the state that emulates something like `$debug_meta.images[].code_name`.

jan-auer replied:
We can also add this at a later stage, this is still G2G

Commit:
Nicer testing infrastructure and tests for linux and mac minidumps

Writing tests with this is slightly nicer.

Commit:
Add Display impl for WStr

Commit:
Add double-ended char iters to WStr

This primarily implements the DoubleEndedIterator to WStrChars and
WStrCharIndices.

It also introduces two new helpers: for constructing a u16 from 2 bytes
and for constructing a char from two u16s.

Finally it optimises the Iter::count and Iter::last implementations of
the char iters.

Commit:
Tiny bit more efficient basename splitting

This was not worth the time spent on it, but was fun.

Commit:
Update readme

untitaker reviewed (approved):
already reviewed this a couple of times

Commit:
Clippy escaped me

Commit:
Fix integration tests

They were applying some pii config to the minidump, but the test was
checking the attachment (the minidump) was not being modified.

jan-auer commented on CHANGELOG.md:
```diff
@@ -4,7 +4,7 @@
 **Features**:
 
-- Add support for scrubbing UTF-16 data in attachments ([#742](https://github.com/getsentry/relay/pull/742))
+- Add support for scrubbing UTF-16 data in attachments ([#742](https://github.com/getsentry/relay/pull/742)) ([#784](https://github.com/getsentry/relay/pull/784))
```
```suggestion
- Add support for scrubbing UTF-16 data in attachments ([#742](https://github.com/getsentry/relay/pull/742), [#784](https://github.com/getsentry/relay/pull/784))
```

jan-auer committed:
meta: Changelog
